File: langgraph_essay_grading/graph.py
"""工作流图定义"""

from langgraph.graph import END, START, StateGraph

from langgraph_essay_grading.nodes import (
    calculate_final_score,
    check_evidence,
    check_expression,
    check_relevance,
    check_structure,
)
from langgraph_essay_grading.routes import route_relevance
from langgraph_essay_grading.state import EssayState


def fan_out(state: EssayState) -> dict:
    """扇出节点：触发三个并行评分。"""
    return {}


# 构建状态图
workflow = StateGraph(EssayState)

# 注册节点
workflow.add_node("check_relevance", check_relevance)
workflow.add_node("fan_out", fan_out)
workflow.add_node("check_evidence", check_evidence)
workflow.add_node("check_structure", check_structure)
workflow.add_node("check_expression", check_expression)
workflow.add_node("calculate_final_score", calculate_final_score)

# 定义边
workflow.add_edge(START, "check_relevance")
workflow.add_conditional_edges(
    "check_relevance",
    route_relevance,
    {
        "fan_out": "fan_out",
        "calculate_final_score": "calculate_final_score",
    },
)
# 扇出：三个评分节点并行执行
workflow.add_edge("fan_out", "check_evidence")
workflow.add_edge("fan_out", "check_structure")
workflow.add_edge("fan_out", "check_expression")
# 汇聚：全部完成后计算总分
workflow.add_edge("check_evidence", "calculate_final_score")
workflow.add_edge("check_structure", "calculate_final_score")
workflow.add_edge("check_expression", "calculate_final_score")
workflow.add_edge("calculate_final_score", END)

# 编译图
graph = workflow.compile()

File: src/workflow/graph.ts
import { END, START, StateGraph } from "@langchain/langgraph";
import { EssayState, type EssayStateType } from "./state";
import {
  calculate_final_score,
  check_content,
  check_depth,
  check_evidence,
  check_expression,
  check_formatting,
  check_novelty,
  check_relevance,
  check_structure,
} from "./nodes";
import { routeRelevance } from "./routes";
import { NODE_NAMES } from "./config";
import type { Mode } from "../lib/mode-storage";

function fanOut(_: EssayStateType): Record<string, never> {
  return {};
}

// 标准模式: 4 维(evidence / structure / expression 并行)
export const standardGraph = new StateGraph(EssayState)
  .addNode(NODE_NAMES.RELEVANCE, check_relevance)
  .addNode(NODE_NAMES.FAN_OUT, fanOut)
  .addNode(NODE_NAMES.EVIDENCE, check_evidence)
  .addNode(NODE_NAMES.STRUCTURE, check_structure)
  .addNode(NODE_NAMES.EXPRESSION, check_expression)
  .addNode(NODE_NAMES.CALCULATE, calculate_final_score)
  .addEdge(START, NODE_NAMES.RELEVANCE)
  .addConditionalEdges(NODE_NAMES.RELEVANCE, routeRelevance, {
    fan_out: NODE_NAMES.FAN_OUT,
    calculate_final_score: NODE_NAMES.CALCULATE,
  })
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.EVIDENCE)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.STRUCTURE)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.EXPRESSION)
  .addEdge(NODE_NAMES.EVIDENCE, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.STRUCTURE, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.EXPRESSION, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.CALCULATE, END)
  .compile();

// 高考模式: 6 维(content / structure / expression / depth / novelty / formatting 并行)
export const gaokaoGraph = new StateGraph(EssayState)
  .addNode(NODE_NAMES.RELEVANCE, check_relevance)
  .addNode(NODE_NAMES.FAN_OUT, fanOut)
  .addNode(NODE_NAMES.CONTENT, check_content)
  .addNode(NODE_NAMES.STRUCTURE, check_structure)
  .addNode(NODE_NAMES.EXPRESSION, check_expression)
  .addNode(NODE_NAMES.DEPTH, check_depth)
  .addNode(NODE_NAMES.NOVELTY, check_novelty)
  .addNode(NODE_NAMES.FORMATTING, check_formatting)
  .addNode(NODE_NAMES.CALCULATE, calculate_final_score)
  .addEdge(START, NODE_NAMES.RELEVANCE)
  .addConditionalEdges(NODE_NAMES.RELEVANCE, routeRelevance, {
    fan_out: NODE_NAMES.FAN_OUT,
    calculate_final_score: NODE_NAMES.CALCULATE,
  })
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.CONTENT)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.STRUCTURE)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.EXPRESSION)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.DEPTH)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.NOVELTY)
  .addEdge(NODE_NAMES.FAN_OUT, NODE_NAMES.FORMATTING)
  .addEdge(NODE_NAMES.CONTENT, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.STRUCTURE, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.EXPRESSION, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.DEPTH, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.NOVELTY, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.FORMATTING, NODE_NAMES.CALCULATE)
  .addEdge(NODE_NAMES.CALCULATE, END)
  .compile();

export function getGraph(mode: Mode) {
  return mode === "gaokao" ? gaokaoGraph : standardGraph;
}

// 向后兼容: 旧代码用 `graph` 导入时默认给 standard
export const graph = standardGraph;
